tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+        // been removed from the group.
+        sender_id: &Vec<u8>,
+        group: &mut Group,
+        messages: Vec<MLSMessage>,
+    ) -> Result<(), ClientError> {
+        let clients = self.clients.borrow();
+        println!("Distributing and processing messages...");
+        // Distribute message to all members.
+        for (index, member_id) in &group.members {
+            println!("Index: {:?}, Id: {:?}", index, member_id);
+            let member = clients.get(member_id).unwrap().borrow();
+            member.receive_messages_for_group(&group.group_id, messages.clone())?;
+        }
+        // Get the current tree and figure out who's still in the group.
+        let sender = clients.get(sender_id).unwrap().borrow();
+        let sender_groups = sender.groups.borrow();
+        let sender_group = sender_groups.get(&group.group_id).unwrap();
+        group.members = sender
+            .get_members_of_group(&group.group_id)?
+            .iter()
+            .map(|(index, cred)| (index.clone(), cred.identity().clone()))
+            .collect();
+        group.public_tree = sender_group.export_ratchet_tree();
+        drop(sender_group);
+        drop(sender_groups);
+        println!("Group members after distribution:");
+        // Check that the group states of all members match.
+        for (index, m_id) in &group.members {
+            println!("Id: {:?}, Index: {:?}", m_id, index);
+            let m = clients.get(m_id).unwrap().borrow();
+            let group_states = m.groups.borrow_mut();
+            // Some group members may not have received their welcome messages yet.
+            if let Some(group_state) = group_states.get(&group.group_id) {
+                assert_eq!(group_state.export_ratchet_tree(), group.public_tree);
+            };
+            drop(group_states);
+        }
+        Ok(())
+    }
+
+    /// Get `number_of_members` new members for the given group `group` for use
+    /// with the `add_members` function.
+    pub fn random_new_members_for_group(
+        &self,
+        group: &Group,
+        number_of_members: usize,
+    ) -> Result<Vec<Vec<u8>>, SetupError> {
+        let clients = self.clients.borrow();
+        if number_of_members + group.members.len() >= clients.len() {
+            return Err(SetupError::NotEnoughClients);
+        }
+        let mut new_member_ids = Vec::new();
+        for _ in 0..number_of_members {
+            let new_member_id = clients
+                .keys()
+                .find(|&client_id| {
+                    (group
+                        .members
+                        .iter()
+                        .find(|&(_, member_id)| client_id == member_id)
+                        .is_none())
+                        && (new_member_ids
+                            .iter()
+                            .find(|&member_id| member_id == client_id)
+                            .is_none())
+                })
+                .unwrap();
```

Comment:
So this panics if there's no client we could add?

Reply:
Well, there really should be one. We check in the beginning if enough clients exist.